unison-share-api/src/Unison/Server/CodebaseServer.hs

```diff
@@ -285,31 +300,52 @@ serveIndex path = do
                   <> " environment variable to the directory where the UI is installed."
           }
 
-serveUI :: Handler () -> FilePath -> Server WebUI
-serveUI tryAuth path _ = tryAuth *> serveIndex path
+serveUI :: FilePath -> Server WebUI
+serveUI path _ = serveIndex path
 
 server ::
   Rt.Runtime Symbol ->
   Codebase IO Symbol Ann ->
   FilePath ->
   Strict.ByteString ->
-  Server AuthedServerAPI
-server rt codebase uiPath token =
+  Server AppAPI
+server rt codebase uiPath expectedToken =
   serveDirectoryWebApp (uiPath </> "static")
-    :<|> ( \token ->
-             serveUI (tryAuth token) uiPath
-               :<|> unisonApi token
-               :<|> serveOpenAPI
-               :<|> Tagged serveDocs
-         )
+    :<|> hoistWithAuth serverAPI expectedToken serveServer
+  where
+    serveServer :: Server ServerAPI
+    serveServer =
+      ( serveUI uiPath
+          :<|> serveUnisonAndDocs rt codebase
+      )
+
+serveUnisonAndDocs :: Rt.Runtime Symbol -> Codebase IO Symbol Ann -> Server UnisonAndDocsAPI
+serveUnisonAndDocs rt codebase = serveUnison codebase rt :<|> serveOpenAPI :<|> Tagged serveDocs
+
+serveDocs :: Application
+serveDocs _ respond = respond $ responseLBS ok200 [plain] docsBS
   where
-    serveDocs _ respond = respond $ responseLBS ok200 [plain] docsBS
-    serveOpenAPI = pure openAPI
+    plain :: (HeaderName, ByteString)
     plain = ("Content-Type", "text/plain")
-    tryAuth = handleAuth token
-    unisonApi t =
-      NamespaceListing.serve (tryAuth t) codebase
-        :<|> NamespaceDetails.serve (tryAuth t) rt codebase
-        :<|> Projects.serve (tryAuth t) codebase
-        :<|> serveDefinitions (tryAuth t) rt codebase
-        :<|> serveFuzzyFind (tryAuth t) codebase
+
+serveOpenAPI :: Handler OpenApi
+serveOpenAPI = pure openAPI
+
+hoistWithAuth :: forall api. HasServer api '[] => Proxy api -> ByteString -> ServerT api Handler -> ServerT (Authed api) Handler
+hoistWithAuth api expectedToken server token = hoistServer @api @Handler @Handler api (\h -> handleAuth expectedToken token *> h) server
+
+serveUnison ::
+  Codebase IO Symbol Ann ->
+  Rt.Runtime Symbol ->
+  Server UnisonAPI
+serveUnison codebase rt =
+  hoistServer (Proxy @UnisonAPI) backendHandler $
+    (\root rel name -> setCacheControl <$> NamespaceListing.serve codebase root rel name)
```

Comment:
Here's where the cache control headers are set. It's a bit of annoying boilerplate to re-apply all the args, but it's not a big deal.